### ISSUE TITLE
User-Defined Version Comparison Methods

### ISSUE THREAD

**Issue by mouyang** [mojohaus/versions-maven-plugin#194]
The current versions plugin only supports 3 comparison methods (maven, numeric, mercury).  It would be very helpful additional comparison methods could be registered with the VersionComparators and used in version number rules.

Here are a few situations where this feature could be useful: 
* Not all libraries support the standard versioning conventions rendering the plugin useless for those libraries.  Some examples include Hibernate and Spring Framework.  This feature would allow comparators to be defined for them.
* Larger companies may have their versioning conventions for their libraries that don't follow the supported methods.
* ComparableVersion in Maven Artifact is currently not supported by the Versions plugin.

In terms of defining comparison methods, I think it would be reasonable to add a new element in the ruleset xml as follows

```xml
<ruleset comparisonMethod="maven"
         xmlns="http://mojo.codehaus.org/versions-maven-plugin/rule/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://mojo.codehaus.org/versions-maven-plugin/rule/2.0.0 http://mojo.codehaus.org/versions-maven-plugin/xsd/rule-2.0.0.xsd">
  <versionComparators>
    <versionComparator name="myVersionComparator" comparatorClass="org.codehaus.mojo.versions.ordering.MyVersionComparator" />
  </versionComparators>
  <rules>
    <rule groupId="*.maven" comparisonMethod="myVersionComparator"/>
  </rules>
</ruleset>
```

This depends on two things currently marked as deprecated: #157 which looks like it's going to stay, and VersionComparators.

I would be willing to try to implement this but I'd like to know if it's even worth pursuing in the first place.

**Comment by khmarbaise** [comment]
The question is what kind of versioning would you like to support, cause the base of all that is Maven itself which has an implementation of version comparison and what kind of problem you are trying to solve..

**Comment by mouyang** [comment]
Here are some examples that are used in my codebase where the plugins give the wrong version

Type 1 - Qualifiers preceded by "." instead of "-" as required by DefaultArtifactVersion

1. Maven Central Link: https://mvnrepository.com/artifact/org.springframework/spring-core
display-dependency-updates output: org.springframework:spring-core ............... 4.2.7.RELEASE -> 2.5.6
1. Maven Central Link: https://mvnrepository.com/artifact/org.hibernate/hibernate-entitymanager
display-dependency-updates output: org.hibernate:hibernate-entitymanager .... 5.0.12.Final -> 3.5.6-Final
1. Maven Central Link: https://mvnrepository.com/artifact/org.mvel/mvel2
display-dependency-updates output:  org.mvel:mvel2 ........................... 2.3.1.Final-<custom_extended_qualifier> -> 2.0.19

The common themes that seem to in play are that 
1. The "latest" is determined to be the latest version that fits format required by DefaultArtifactVersion. 
1. When examining the examples above, some version format follow the required format and some do not.

Type 2 - Custom Qualifiers

This example follows the DefaultArtifactVersion format, but the qualifier is compared lexicographically.  A user-defined comparator would allow the qualifier to be compared correctly.

1. Maven Central Link: https://mvnrepository.com/artifact/org.asciidoctor/asciidoctorj-pdf
display-dependency-updates output: org.asciidoctor:asciidoctorj-pdf ..... 1.5.0-alpha.14 -> 1.5.0-alpha.9

It also seems like there's talks to re-do the versioning in Maven altogether so it might be a good idea to have a way to hook into that (https://cwiki.apache.org/confluence/display/MAVENOLD/Versioning) although I don't know if any progress has been made on that.

**Comment by mouyang** [comment]
Here are some examples that are used in my codebase where the plugins give the wrong version

Case 1 - Qualifiers preceded by "." instead of "-" as required by DefaultArtifactVersion

1. Maven Central Link: https://mvnrepository.com/artifact/org.springframework/spring-core
display-dependency-updates output: org.springframework:spring-core ............... 4.2.7.RELEASE -> 2.5.6
1. Maven Central Link: https://mvnrepository.com/artifact/org.hibernate/hibernate-entitymanager
display-dependency-updates output: org.hibernate:hibernate-entitymanager .... 5.0.12.Final -> 3.5.6-Final

The common themes that seem to in play are that 
1. The "latest" is determined to be the latest version that fits format required by DefaultArtifactVersion. 
 For these particular libraries, I would argue the plugin would be not usable.
1. When examining the examples above, some version format follows the required format and some do not.

Case 2 - Custom Qualifiers

This example follows the DefaultArtifactVersion format, but the qualifier is compared lexicographically.  A user-defined comparator would allow the qualifier to be compared correctly.  The qualifier is effectively freeform and project-specific, therefore something needs to be put in place to handle it.

1. Maven Central Link: https://mvnrepository.com/artifact/org.asciidoctor/asciidoctorj-pdf
display-dependency-updates output: org.asciidoctor:asciidoctorj-pdf ..... 1.5.0-alpha.14 -> 1.5.0-alpha.9

Case 3 - "New" way of doing Versioning in Maven
It also seems like there's talks to re-do the versioning in Maven altogether so it might be a good idea to have a way to hook into that (https://cwiki.apache.org/confluence/display/MAVENOLD/Versioning) although I don't know if any progress has been made on that.

**Comment by mouyang** [comment]
Hello.  I wanted to provide an update on my progress on this issue.  I failed to provide details about my environment.  It happened in 2.3 and upgrading to 2.5 seems to fix the issue.  My team was stuck on Maven 2 and so I did not do this upgrade until recently.

I ended up doing the refactoring before realizing this anyways.  I do think there's still some value in this feature, although not as much as before, just in case a artifact's versioning scheme doesn't mesh with the default provided.
